### PR TITLE
Fixes #303 - constraint issue (create before delete) with mergeNodes

### LINKED ISSUE #303: constraint issue (create before delete) with mergeNodes
Failed to invoke procedure `apoc.refactor.mergeNodes`: Caused by: org.neo4j.graphdb.ConstraintViolationException: Node 1250 already exists with label Organization and property "name"=[INVESCO, LTD.]

we need to delete the node first (or at least the key property) before creating a new one

### PR DESCRIPTION
**Pull request by DanielBerton**

### CODE REVIEW

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 1]
```diff
@@ -299,16 +299,30 @@ public void categorize(
     }
 
     private Node mergeNodes(Node source, Node target, boolean delete) {
-        copyRelationships(source, copyProperties(source, copyLabels(source, target)), delete);
-        if (delete) source.delete();
+        Map<String, Object> propTemp = new HashMap<>();
+        List<String> keys = new ArrayList<>();
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: why do we need keys?

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 2]
```diff
@@ -299,16 +299,30 @@ public void categorize(
     }
 
     private Node mergeNodes(Node source, Node target, boolean delete) {
-        copyRelationships(source, copyProperties(source, copyLabels(source, target)), delete);
-        if (delete) source.delete();
+        Map<String, Object> propTemp = new HashMap<>();
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: can we find a better name? why not just Map<String, Object> properties = source.getAllProperties()

Didn't we say that we just store/keep the existing properties and then set the new ones after deletion of the source node? Not sure why we have to remove the properties individually?

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 3]
```diff
@@ -299,16 +299,30 @@ public void categorize(
...
+            propTemp.put(prop.getKey(), prop.getValue());
+            keys.add(prop.getKey());
+            source.removeProperty(prop.getKey());
+        }
+        copyRelationships(source, copyLabels(source, target), delete);
+        setProperty(propTemp, keys, target);
         return target;
     }
 
+    private void setProperty(Map<String, Object> propTemp, List<String> keys, Node target) {
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: isn't that the same as copy properties? only from a map source? Best name it the same and have it return the target node too.

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 4]
```diff
@@ -299,16 +299,30 @@ public void categorize(
...
+            target.setProperty(keys.get(i), propTemp.get(keys.get(i)));
+        }
+    }
+
     private Node copyRelationships(Node source, Node target, boolean delete) {
         for (Relationship rel : source.getRelationships()) {
             copyRelationship(rel, source, target);
             if (delete) rel.delete();
         }
+        if (delete) source.delete();
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: will this break any other code that uses this? I think the original contract for this method was that rels are deleted but the nodes are left alone.
so I'd rather move this out of the method back into mergeNodes where it originally was.

File: src/main/java/apoc/refactor/GraphRefactoring.java (hunk 3 above)
Comment (DanielBerton) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: @jexp so we remane it `copyProperties` with this signature `private Node copyProperties(Map<String, Object> properties, Node target)` right?

File: src/main/java/apoc/refactor/GraphRefactoring.java (hunk 4 above)
Comment (DanielBerton) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: @jexp ok so we move back the delete into `mergeNodes ` before the call of ` copyProperties`  method

File: src/main/java/apoc/refactor/GraphRefactoring.java (hunk 3 above)
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: yes

File: src/main/java/apoc/refactor/GraphRefactoring.java (hunk 4 above)
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: yes

**Comment by DanielBerton** [comment]
@jexp change requests fixed

**Comment by jexp** [comment]
Looks great, thanks a lot.

**Pull request merged**